let users point a bazooka a their foot wtih SetupKwargs plugins

cburroughs opened the pull request:
Preciously one could write a plugin and with `_allow_banned_keys` set something like `install_requires`.  The new and even more dangerous `_overwrite_banned_keys` allows one to exclusively use values calculated in their plugin.

The motivating use case is a plugin that builds a wheel with all dependencies pinned to the value from lockfile.

ref #21252

cognifloyd commented:
> The motivating use case is a plugin that builds a wheel with all dependencies pinned to the value from lockfile.

Fascinating. I would love to see this plugin and maybe use it. :)

cburroughs commented:
The (somewhat hacky) critical plugin section is:

```
    if pin_external_dependencies: # <---
        requirements_pex = await Get(
            Pex,
            PexFromTargetsRequest(
                addresses=[request.target.address],
                output_filename="requirements.pex",
                internal_only=True,
                include_source_files=False,
                additional_args=("--include-tools",),
            ),
        )
        result = await Get(
            ProcessResult,
            PexProcess(
                requirements_pex,
                extra_env={
                    "PEX_TOOLS": "1",
                },
                argv=("repository", "info"),
                description=f"Pinning external dependencies for {request.target.address}",
            ),
        )
        out = result.stdout.decode("utf-8").strip()
        repository_info = out.split("\n")
        requirements = sorted(
            [
                f"{tup[0].strip()}=={tup[1].strip()}"
                for tup in [line.split(" ") for line in repository_info]
            ]
        )
        final_requirements = []
        for requirement in requirements:
            if any(
                [
                    requirement.startswith(ignored)
                    for ignored in ignore_pinned_external_dependencies
                ]
            ):
                continue
            final_requirements.append(requirement)
        allow_banned_keys = True
        # NOTE: These get "combined" with the original unpinned requirements,
        # see https://github.com/pantsbuild/pants/issues/21252
        extra_kwargs["install_requires"] = final_requirements
```

Where the use looks something like:
```
python_distribution(
    name="super-cool-name",
    provides=python_artifact(
        name="super-cool-name",
        corpname_pin_3rdparty_requirements=True,
        corpname_ignore_pinned_3rdparty_requirements=[
            "tensorflow",
            "tensorflow-macos",
            "tensorflow-metal",
        ],
    ),
    sdist=False,
    dependencies=[
        ":all_python",
    ],
)
```
